### PR TITLE
Refactor catching of OSErrors when finding default temp directories, and improve the naming of fallback temp directories for Linux and Mac platforms

### PR DESCRIPTION
**Pull request by tanchihpin0517**
Hi,
I add two checks for read & write permission of the root temp directory.
I also modify the naming rule of alternative temp directory for ~Linux~ Linux and Mac platform with "music21-$UID" (which is "music21-$(8-digit hashed code)" originally). It can avoid music21 creating too many directories while using the alternative temp directory.

### CODE REVIEW

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:510]: This is too complex for a doctest, which is supposed to show users how to use the API. But I understand you wanted to doctor it so that it would pass on CI running Linux. Unfortunately we need to fake the doctest a little bit with #_DOCS_HIDE or #_DOCS_SHOW to hide and show lines to run or not run (see examples elsewhere in this file), and then move this test to a unit test. You can add a decorator for skipUnless so that your test only runs on Linux platforms.

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:532]: I like this, thanks. 👍🏻

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:541]: Ah, is this where we were creating too many directories? Each time User X lacks write access to the root temp dir, we catch `OSError` and then attempt to create more directories with new hashes each time? Makes sense to change this. But is there a reason you didn't want to use `prefix=` like below? I just feel like we should have both branches behave the same way, either construct the tempdir inline, or construct a path and then mkdir() after.

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
Thanks, this looks like a nice improvement. Left some comments to move this forward. Let me know if you have any questions (or if I've misunderstood something).

File: music21/environment.py
Comment (tanchihpin0517) [reply music21/environment.py:539]: Yes. I didn't notice there are the common tools in this project. I'll fix it.

File: music21/environment.py
Comment (tanchihpin0517) [reply music21/environment.py:510]: Got it. But I'm not familiar with doc tests and unit tests. It may take some time :P

File: music21/environment.py
Comment (tanchihpin0517) [reply music21/environment.py:541]: > Ah, is this where we were creating too many directories? Each time User X lacks write access to the root temp dir, we catch OSError and then attempt to create more directories with new hashes each time?

Yes. It creates a new directory for each different python session.

> But is there a reason you didn't want to use prefix= like below? I just feel like we should have both branches behave the same way, either construct the tempdir inline, or construct a path and then mkdir() after.

I have tried to make code here more elegant. But tempfile doesn't provide suitable API to make both branches behave in the same way. Maybe I can change it to
```
...
dir_path = pathlib.Path(tempfile.gettempdir()) / f'music21-{uid}'
dir_path.mkdir()
...
```
and remove the if block
```
...
if not dir_path.exists():
    dir_path.mkdir()
...
```

File: music21/environment.py
Comment (jacobtylerwalls) [reply music21/environment.py:510]: No worries! The good news is that the doctest passes with your feature without having to alter it, so I think we should just leave this doctest as is. I verified [here--not-real PR for testing only](https://github.com/jacobtylerwalls/music21/pull/29).

Then finding a way to test this works in a unit test -- probably better to mock the results of `os.access()` if possible than change read/write permissions of `tmp/`. Maybe you can look into [unittest mocking](https://docs.python.org/3/library/unittest.mock.html)?

File: music21/environment.py
Comment (jacobtylerwalls) [reply music21/environment.py:541]: Sure. I was just wondering why we weren't using `tempfile.mkdtemp()`. I know `NamedTemporaryDirectory` is preferred over mkdtemp, but I wasn't sure why `gettempdir()` and then `mkdir()` is superior to just mkdtemp. Honestly just wondering if you had a reason or if I'm missing something.

File: music21/environment.py
Comment (jacobtylerwalls) [reply music21/environment.py:510]: I took the other approach in 10f097274e of trying to fiddle with permissions in a unittest when you and I were working in this area last, but because I was trying to be cautious about fiddling with people's systems, the test doesn't run on CI, and so we didn't notice that it needs updating with your changes:

https://github.com/cuthbertLab/music21/blob/10f097274e82199132f50073d6cd89ba2ca85411/music21/environment.py#L1647

So I guess I'm open to both possibilities, you can try to fiddle with permissions, but honestly it's probably better to just mock. But we should update the test I just pointed to.

Comment (jacobtylerwalls) [reply music21/environment.py:510]: As in, the test passes on my Mac because it's not Linux, but if this test ever runs on a Linux system where there is read/write access to tmp/ (apparently not on GitHub Actions) it will fail because `getpass.getuser()` != `os.getuid()`.

Comment (tanchihpin0517) [reply music21/environment.py:510]: I check [10f0972](https://github.com/cuthbertLab/music21/commit/10f097274e82199132f50073d6cd89ba2ca85411). I agree that it needs to be update. But I found these two part is used to deal with the same problem. It looks a little bit redundant. Should we integrate them into a single block? I mean remove the exception handling in `getTempFile()`, and only catch it in `getDefaultRootTempDir()`. Cause `getTempFile()` use `getDefaultRootTempDir()` to get the temp directory, we just need to make sure that `getDefaultRootTempDir()` provides a accessible and permission-allowed directory. Then we only need to maintain a single code if a new problem occurs in the future,

File: music21/environment.py
Comment (tanchihpin0517) [reply music21/environment.py:541]: The reason is `tempfile.mkdtemp()` create a directory with random name (even called with `prefix`). It can't be chased in different python session. If program can't get the last-created temp directory, then it have to create a new one. There are two solution as I know. The first one is storing the name of temp directory somewhere in the disk, and the second one is using a fixed name. I prefer the second one so I use `gettempdir()+suffix_name` and then `mkdir()`.

Comment (jacobtylerwalls) [reply music21/environment.py:541]: Thanks, that helps me understand! All good.

**Comment by mscuthbert** [comment]
Our temp dir isn't supposed to be extremely temporary, which is what TempDir tries to do.  We'd like it to persist until the system decides it needs to reclaim the space.  

Agree w/ Jacob that these OS things shouldn't be in a doctest, better as a unittest that only runs on `if common.getPlatform() == 'nix'` -- and if possible, refactor out anything specific to unix distribution to a separate function/method.  As much as possible try to be explicit so people know what we are doing, so for instance instead of `'music21-{uid}'` use `music21-userid-{uid}`

File: music21/environment.py
Comment (jacobtylerwalls) [reply music21/environment.py:510]: Yes, that sounds great, let's simplify here and just have the permission checking in getDefaultRootTempDir().

**Comment by tanchihpin0517** [comment]
Hi, I have finished the modification.
I follow your recommendations, using common tools to check platform and simplifying doc test by moving it to unit test. I also integrate the error checking in `getTempFile` into `getDefaultRootTempDir` and remove the unneeded unit test for `getTempFile`.
I push a new temporary commit to keep the edit history. I will integrate(with git rebase) all commits into a single one before merging.

**Comment by jacobtylerwalls** [comment]
> I push a new temporary commit to keep the edit history. I will integrate(with git rebase) all commits into a single one before merging. But I'm not sure if it's correct cause I see the "Changes reuested" still marked. Should I integrate the new commit into the old one directly?

It's all good. We will squash or rebase when merging. And "requested changes" should go away when I can provide a second review in the next few days. Thanks for the updates!

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:1606]:
```suggestion
            # Restore original permissions and original path
```

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:1574]: Love that this test runs on GitHub actions now. But now it won't run on Macs. Can we add a different assertion for Macs inside this test so that we can still only skip this test if windows? Hoping it's just a matter of checking platform and then making a different assertion, not causing you to rewrite the permissions fiddling.

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:439]: This is private, so not a big deal, but it's usually easier to annotate the type than to spell it out in English.
```suggestion
    def _checkAccessibility(self, path: Union[str, pathlib.Path]) -> bool:
        '''
        Return True if the path exists, and is readable and writable.
        '''
```

**Review by jacobtylerwalls — APPROVED** [review]
This is great to have both the behavior change as well as a better separation of concerns between `getTempFile()` and `getDefaultRootTempDir()`. Just some minor changes at this point. 

Windows test suite looks good: https://github.com/jacobtylerwalls/music21/pull/31 (existing failures known: musedata converter failing on Windows)

Michael also might have some comments. Thanks again.

File: music21/environment.py
Comment (tanchihpin0517) [reply music21/environment.py:1574]: This test fail on Mac because the permissions fiddling behaves differently between Unix and other platforms (Mac and Windows). Should we include Mac into the routine of permissions fiddling same as Unix? Or just keep the current one?

Comment (jacobtylerwalls) [reply music21/environment.py:1574]: I get this result if I run the test as written:
```
AssertionError: 'music21-userid-501' not found in '/var/folders/sd/zw9wtxl552j6wn4yvt1nhg5r0000gn/T/music21-2_cvthvx'
```
Can't we just assert something else if system.platform() == 'darwin'? Or is your failure different?

Comment (jacobtylerwalls) [reply music21/environment.py:1574]: Also, `os.getuid()` is available on Mac, because when Python docs say UNIX they include Mac.

```
>>> import os
>>> os.getuid()
501
```

Comment (tanchihpin0517) [reply music21/environment.py:1574]: >　Can't we just assert something else if system.platform() == 'darwin'?

Yeah. That's what I mean for "tinny modification". OK. I will do it.

**Comment by tanchihpin0517** [comment]
Done. Ready for review.

File: music21/environment.py
Comment (jacobtylerwalls) [reply music21/environment.py:1574]: I'm still confused -- I thought you were going to leave the behavior alone and just change the test to assert the getUid() version for Linux and the 8-hashed characters for Mac. But you left the assertions alone, widened the test to run on Macs also, and then changed the behavior to use getUid() for Linux and Mac. I think that's fine, it's just not what I expected you to do based on what you said. If we keep it this way -- again, fine with me -- then we should update the docs above, where you say only Linux platforms have the getUid() version and other platforms have 8 hashed characters. Thanks.

Comment (jacobtylerwalls) [reply music21/environment.py:1574]: > and if possible, refactor out anything specific to unix distribution to a separate function/method

I agree with Michael's impulse to factor any Linux workarounds out so that they are easier to maintain (or ignore). But @tanchihpin0517 identified some two methods that weren't cleanly separated, so upon refactoring there really isn't much that's Linux specific anymore. What's different is that Windows doesn't have `getuid()`. Mac users are less likely than Linux users to hit permission errors, but we should still try to catch the permission errors and put user-friendly names on the fallback dirs we create. Just thinking out loud, not recommending any changes.

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:522]: Linux and Mac

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:529]:
```suggestion
        Note the temporary directory may be different in each python session.
```

File: music21/environment.py
Comment (jacobtylerwalls) [inline music21/environment.py:1609]: Great catch -- because the second fallback if we still get OSError trying to create the fallback might just be /tmp, and we don't want to try to delete that. Maybe add a comment to that effect? We want to make this easy for someone five years from now to see what's going on.

File: music21/environment.py
Comment (tanchihpin0517) [reply music21/environment.py:1574]: > I'm still confused -- I thought you were going to leave the behavior alone and just change the test to assert the getUid() version for Linux and the 8-hashed characters for Mac.

It's my fault. I accepted your recommendation, doing permission fiddling on Unix (including Mac) rather than only on Linux, but didn't change the doc. I think that's why you feel confused.

**Comment by tanchihpin0517** [comment]
Update here:

After discussing, we use fix-named alternative temp directory on Unix-like platforms (Linux and Mac), and use the one named with 8-digit random hashed codes on other platforms (Windows). Thanks for reminding.

**Comment by tanchihpin0517** [comment]
Hi. This PR is suspended here for some time. I wrap up commits into a single one. This PR is ready for merging now.
Please remind me if there are something I miss. Thanks.

**Comment by jacobtylerwalls** [comment]
Hi and thanks for the note. Nothing for you to do at the moment, we just need two reviewers to look at it. For FOSS projects with a significant volume of PRs patience is appreciated. Two weeks is actually a pretty quick turnaround for a FOSS project.

**Comment by mscuthbert** [comment]
Hi @tanchihpin0517  -- just a note that force-pushing makes it so that reviewers have to start over from the beginning in reviewing the PR, so my "changes since last review" is absent.  This is part of what has slowed the review.  Looking now.

File: music21/environment.py
Comment (mscuthbert) [inline music21/environment.py:435]: path typing should include 'None' as a possibility, since rootTempDir can be None, and otherwise the isinstance check does not make sense below.

File: music21/environment.py
Comment (mscuthbert) [inline music21/environment.py:1612]: because this is in `finally` code, cannot assume that `newTempDir` is defined at this point.  Define it as None outside the try and then test to see if it is not None first before running samefile.  Should also check that it is not the same file as 'directoryScratch' or oldScratchDir before running.

**Review by mscuthbert — CHANGES_REQUESTED** [review]
The code for the examples is fine, but the test code is too dangerous to approve.  See comments below.

File: music21/environment.py
Comment (jacobtylerwalls) [reply music21/environment.py:435]: @tanchihpin0517 this was my fault, sorry, you just committed my suggestion.

```suggestion
    def _checkAccessibility(self, path: Optional[Union[str, pathlib.Path]]) -> bool:
```

**Comment by mscuthbert** [comment]
Sorry -- I;m not seeing my main review comment appearing here -- it's about the shutil `rmtree` -- I think it needs to be removed and only the directories created during the creation deleted, otherwise the testcode could potentially wipe out a whole filesystem if not careful.

**Comment by mscuthbert** [comment]
I can't find where I wrote an extensive comment, but basically I'm concerned that test code is using rmtree -- it is safe as it currently is, but the point of test code is that someone should be able to tinker with the underlying code in the future and run tests to see if it works.  What if someone changes the code or a later OS update changes returning getTempDir to return '/' for privacy?  Then the test code could potentially destroy an entire hard drive.  The test code in this case should individually remove whatever files it created one-by-one.

**Comment by mscuthbert** [comment]
Converting to draft for safety to make sure it's not merged until that's fixed.

**Comment by tanchihpin0517** [comment]
Sorry. I didn't notice recent updates here.
I make two changes:
1. make sure codes have been run in the `try` block before do something in the `final` block.
2. delete `newTempDir` only if it's empty. It can guarantee that the action of deleting is safe.

**Comment by mscuthbert** [comment]
Looks GREAT! to me.  Thank you.  Merging.

**Pull request merged**
